### DISCUSSION TITLE
Updating an atom unmounts and mounts component inside the Styled component

### DISCUSSION THREAD

**Discussion by GoaKoala** [pmndrs/jotai#2606] [Bug report] [answered]
When you update an atom inside a page, components in the page that are placed inside a styled component gets unmounted an mounted again. (NextJs | Typescript | MUI) 

I created a test atom and inside page.tsx(app/page.tsx) I am updating the atom on an interval. Everytime atom is updated, component (StyledComponentContent) inside a styled component (here it is a styled AppBar mui component but any styled component has the same effect) get unmounted and mounted again on every update. 

Component inside a non styled component (NonStyledComponent.tsx & NonStyledComponentContent.tsx) does not show the same behaviour. 

I believe this is an unintended behaviour.

Start the Preview Externally in the CodeSandbox and view the console to see the constant unmounted/mounted logs coming from the component(StyledComponentContent) inside the styled AppBar.
[CodeSandbox](https://codesandbox.io/p/github/GoaKoala/jotai-styled-update-error/master?import=true&layout=%257B%2522sidebarPanel%2522%253A%2522GIT%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clx1uwa2j00063j6i91sd07dk%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clx1uwa2j00023j6i0hi3i94n%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clx1uwa2j00043j6izjrpn2zw%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clx1uwa2j00053j6ijendz45v%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clx1uwa2j00023j6i0hi3i94n%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx1uwa2j00013j6ixrj219io%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522clx1uwa2j00023j6i0hi3i94n%2522%252C%2522activeTabId%2522%253A%2522clx1uwa2j00013j6ixrj219io%2522%257D%252C%2522clx1uwa2j00053j6ijendz45v%2522%253A%257B%2522id%2522%253A%2522clx1uwa2j00053j6ijendz45v%2522%252C%2522activeTabId%2522%253A%2522clx1v2n6n00h93j6ijhkjjblf%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clx1v2n6n00h93j6ijhkjjblf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clx1uwa2j00043j6izjrpn2zw%2522%253A%257B%2522id%2522%253A%2522clx1uwa2j00043j6izjrpn2zw%2522%252C%2522activeTabId%2522%253A%2522clx1uwqid00613j6i9q2jz66m%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx1uwa2j00033j6irohu9atl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clx1uwadb000wdljpbl0a21ou%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clx1uwqid00613j6i9q2jz66m%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

[GitHub](https://github.com/GoaKoala/jotai-styled-update-error)

**Comment by dai-shi** [comment]
Thanks for reporting. Can anyone dig into it and investigate more? Thanks!

**Comment by rothsandro** [comment, marked as answer]
You create the styled component `AppBarStyled` inside another React component but they should always be [created outside of the React component](https://styled-components.com/docs/basics#define-styled-components-outside-of-the-render-method) (or the render method in case of class components). Your current solution creates a new React component on every render, that's why all child components are destroyed and re-created (unmounted + mounted). It's not related to Jotai but can be reproduced with `useState()` as well.

Just create the styled component outside:

```tsx
const StyledComponent = () => {
  // WRONG 
  // const AppBarStyled = styled(AppBar)(({ theme }) => ({ }));

  return (
    <AppBarStyled>
      <StyledComponentContent></StyledComponentContent>
    </AppBarStyled>
  );
};

// CORRECT
const AppBarStyled = styled(AppBar)(({ theme }) => ({ }));
```

**Reply by GoaKoala** [reply]
I see. Thanks for the quick reply.
